oauth2_provider/views.py

```diff
@@ -22,8 +23,103 @@
 
 
+class OIDCTokenMixin(OAuth2AccessTokenMixin):
+
+    # pylint: disable=super-on-old-class
+    def access_token_response_data(self, access_token, response_type=None):
+        """
+        Return `access_token` fields for OAuth2, and add `id_token` fields for
+        OpenID Connect according to the `access_token` scope.
+
+        """
+        # Clear the scope for requests that do not use OpenID Connect.
+        # Scopes for pure OAuth2 request are currently not supported.
+        scope = constants.DEFAULT_SCOPE
+
+        extra_data = {}
+
+        # Add OpenID Connect `id_token` if requested.
+        #
+        # TODO: Unfortunately because of how django-oauth2-provider implements
+        # scopes, we cannot check if `openid` is the first scope to be
+        # requested, as required by OpenID Connect specification.
+
+        if not provider.scope.check(constants.OPEN_ID_SCOPE, access_token.scope):
+            if response_type and 'id_token' in response_type.split():
+                # http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest:
+                #   If the openid scope value is not present, the behavior is entirely unspecified.
+                #
+                # For our purposes, we'll treat this as a plain OAuth2 request and fall
+                # back to the default behavior, *unless* the 'id_token' scope is present,
+                # which is OIDC-specific, in which case we assume that the client has made
+                # an error.
+                raise OAuthError({'error': 'invalid_scope'})
+            return super(OIDCTokenMixin, self).access_token_response_data(access_token, response_type)
+
+        if response_type and 'code' not in response_type and 'id_token' not in response_type:
+            # http://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth:
+            #   While OAuth 2.0 also defines the token Response Type value for the Implicit Flow,
+            #   OpenID Connect does not use this Response Type, since no ID Token would be returned.
+            #
+            # A request with the 'openid' scope that does not use either 'code' or
+            # 'id_token' response types is invalid under OIDC.
+            raise OAuthError({
+                u'error': u'invalid_request',
+                u'error_description': u'response_type "{}" is invalid when using scope "openid"'.format(response_type)
+            })
+
+        id_token = self.get_id_token(access_token)
+        extra_data['id_token'] = self.encode_id_token(id_token)
+        scope = provider.scope.to_int(*id_token.scopes)
+
+        # Update the token scope, so it includes only authorized values.
+        access_token.scope = scope
+        access_token.save()
+
+        # Get the main fields for OAuth2 response.
+        response_data = super(OIDCTokenMixin, self).access_token_response_data(access_token, response_type)
+        if response_type and not response_type.endswith(' token'):
```

Comment:
hm...oidc docs imply 'token' will always come after 'id_token' but per rfc6749 the order should not matter, so i will fix this to be order-neutral.